### PR TITLE
[New Algorithm] - Bell Numbers

### PR DESCRIPTION
**Pull request by Tauseef-Hilal**
### Describe your change:
New algorithm for generating Bell numbers.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

File: maths/bell_numbers.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+Example:
+To calculate the Bell numbers for sets of lengths from 0 to 5:
+
+>>> import bell_numbers
+>>> bell_numbers.bell_numbers(5)
+[1, 1, 2, 5, 15, 52]
+"""
+
+
+def bell_numbers(n: int) -> list[int]:
```
[the remark this answers is not in the document]
Comment (Tauseef-Hilal) [reply maths/bell_numbers.py:16]: max_set_length

File: maths/bell_numbers.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    bell[0] = 1
+
+    for i in range(1, n + 1):
+        for j in range(i):
+            bell[i] += _binomial_coefficient(i - 1, j) * bell[j]
+
+    return bell
+
+
+def _binomial_coefficient(n: int, k: int) -> int:
```
[the remark this answers is not in the document]
Comment (Tauseef-Hilal) [reply maths/bell_numbers.py:47]: total_elements, elements_to_choose

File: maths/bell_numbers.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    bell[0] = 1
+
+    for i in range(1, max_set_length + 1):
+        for j in range(i):
+            bell[i] += _binomial_coefficient(i - 1, j) * bell[j]
+
+    return bell
+
+
+def _binomial_coefficient(total_elements, elements_to_choose):
```
[the remark this answers is not in the document]
Comment (Tauseef-Hilal) [reply maths/bell_numbers.py]: done

**Comment by Tauseef-Hilal** [comment]
@algorithms-keeper review

**Comment by Tauseef-Hilal** [comment]
Need review!

**Review by ChrisO345 — APPROVED** [review]

**Pull request merged**